Feature/anonimizer

idelcano opened the pull request:
### :pushpin: References
https://app.clickup.com/t/kb6m0g
### :memo: Implementation

The d2-cloner creates and execute an .sql file with all the queries to anonimoze datasets/trackers and programs.

To anonymize dataelements you must provide a list of valid dataelements (dataelements in the dep programs, valuetype text or long_text...), same way by trackedentityattributes.

to delete (organisation_unit valuetypes)/randomize(org unit attached to program, tei..) the orgunits you must set 
      "anonymizeOrgUnit": true
anonymizePhone, anonymizeMail and anonymizeCoordinates are true by default.

if you select anonymizePhone or email the dataelements in the program(tracker and event) of that types will be removed, we have a really low number of that types and the phone could be encrypted when we solved the confidential problem  and we can't update it by sql.
For the datasets they will be randomized.

Although it might be easy to add randomized emails. 

We could add:
randomize provided dataelements that have optionset, (for example we have some optionsets with a list of countries).
filter anonymizer by org unit and org unit descendants like with the deletion feature.


### :fire: Notes for the reviewer
Example of preproccess in the config file:
Anonymize all ntd programs (delete phone coordinates and email by default and change orgunits if you add "anonymizeOrgUnit": true,):

    {
      "selectDepartament": "NTD",
      "selectMetadataType": ["trackerPrograms","eventPrograms"],
      "anonymizePhone": true,
      "anonymizeMail": true,
      "anonymizeOrgUnit": true,
      "anonymizeCoordinates": true,
      "action": "anonymizeData"
    },
    or if you want randomize too some dataelements with text or long_text valuetype:
    
    {
      "selectDepartament": "GMP_ENTO",
      "selectMetadataType": ["eventPrograms"],
      "selectDataElements": [
          "v7YkIQwsbjV",
          "G3GQUnbk6Y2"
      ],
      "anonymizeOrgUnit": true,
      "action": "anonymizeData"
    }
    example with trackedentityattributes:
    
    {
      "selectDepartament": "NTD",
      "selectMetadataType": ["trackerPrograms"],
      "anonymizePhone": true,
      "anonymizeMail": true,
      "anonymizeOrgUnit": true,
      "anonymizeCoordinates": true,
      "selectTrackedEntityAttributes": ["oTvXfEywjT3", "n8E6WIyAwcC", "DwZNiXy5Daz", "FHw1NKy0PWY", "eQtZaLIO3XU",
        "na3ZJRtjpGH", "HkBG3DVELBM", "sKBh0kazOCk", "AAkZm4ZxFw7", "ENRjVGxVL6l", "aBaYLJryaMr", "iy884aJfYTc"],
      "action": "anonymizeData"
    },
    
Randomize orgunits for each datavalue for the given hwf datasets:

    {
      "selectDepartament": "HWF",
      "selectMetadataType": ["datasets"],
      "selectDatasets": [
          "Tu81BTLUuCT",
          "m5MiTPdlK17"
      ],
      "anonymizeOrgUnit": true,
      "action": "anonymizeData"
    }
    
    
   Example of entire config file without creedentials:
   
```
   {
	"local_type": "d2-docker",
	"local_docker_image": "",
	"local_docker_port": 8081,
	"backups_dir": "",
	"backup_name": "",
	"server_dir_local": "",
	"server_dir_remote": "",
	"post_clone_scripts_dir": "",
	"post_process_import_dir": "",
	"pre_sql_dir": "/tmp/presql",
	"hostname_remote": "",
	"db_remote": "",
	"war_remote": "",
	"api_local": {
		"url": "http://localhost:8081",
		"username": "",
		"password": ""
	},
	"departments": {
		"NTD": {
			"trackerPrograms": [
				"uGCVVme3C4e",
				"NHy39qEl1bc",
				"KHDErSOFmy5",
				"dIr4OtNcvMw",
				"HnA4SUOaTij",
				"j6AF0BiiKUq"
			],
			"eventPrograms": [
				"vFcxda6dJyT",
				"RS6DH7eJJMZ",
				"i5JSf4ffFl2",
				"gxXvKE74vfQ",
				"AODL0NZqmGx",
				"MECZlaC9Ef6",
				"JOwDQ193x2T",
				"ela4dpipbEm",
				"fQTmlJy1rLT",
				"TFTLPseQEgi",
				"uYKG1KOUIy1",
				"h1KVXwrpKKf",
				"dBAcTWcDpbO",
				"NCGFNCiONAq",
				"a7ygv0UrrSK",
				"AOGxHgdhbYD",
				"k3pL2hPUqQi",
				"Jd8gnEIt8uT",
				"nk0PS9gG8dy",
				"EwH68ZzyDRk",
				"Mvj9zd9l5g2",
				"NVUlJzIakuO",
				"IPEvFBRqJjW",
				"w9hSFsNr3Vh",
				"XQl1AKLPgy0",
				"pvYIJL19r6G",
				"qdlYvVgUjU7",
				"wlWD2hWVghi",
				"z6skqIipb47",
				"t1Yjgl9MeOE",
				"CVh71SDylDD",
				"M5pXkRoE62G",
				"K45hjkzs0WX"
			],
			"dataSets": [
				"NePI9W0zSH5",
				"UFlGxXoBrLN",
				"ZO0DQJePIEH",
				"Psf9CW4BSHM",
				"Uc3j0vpsfSB",
				"LdQ3CckNpeQ",
				"tnek2LjfuIm",
				"zna8KfLMXn4",
				"XBgvNrxpcDC",
				"WHPEpoVDFFv",
				"SAV16xEdCZW",
				"AAYgHGENgbF",
				"ZhvE3MziQkC",
				"yOf6XrrS3uy",
				"NKWbkXyfO5F",
				"p0NhuIUoeST",
				"oVxjBKA1Yzu",
				"deKCGAGoEHz",
				"S1UMweeoPsi",
				"fdBM4sWSuPR",
				"SHw2zOysJ1R",
				"s3iaozBY0dv",
				"JP4bMwvJ6oU",
				"U5ejGQdX4Ih"
			]
		},
		"CSY": {
			"trackerPrograms": [
				"YGa3BmrwwiU"
			],
			"eventPrograms": [
				"auqdJ66DqAT",
				"BslciOZ3Rvs",
				"gYhhLQlqduE",
				"nmXOLkibWJc"
			]
		},
		"GHP": {
			"eventPrograms": [
				"azxjVmQLicj",
				"cTzRXZGNvqz"
			],
			"dataSets": [
				"jfawDJZ5fOX",
				"ITKN1N2pPHp",
				"KJbR0F9jUbf"
			]
		},
		"GMP_ENTO": {
			"eventPrograms": [
				"G9hvxFI8AYC",
				"FUzFm6UEmRn",
				"Rw3oD4ExD8U"
			]
		},
		"GMP_EPI": {
			"dataSets": [
				"bNJzkNJv0Yg",
				"qDd74Hfk2Dc",
				"qNtxTrp56wV",
				"Cm6vhZilyPa",
				"iENTl91i0Hg",
				"O34y2Kyxx6P",
				"ctDGeFX3fNZ",
				"tZyUewB5zBf",
				"VEM58nY22sO",
				"VTVvoKniNhi",
				"zyZLeSbJxqR",
				"w9sYDdAm1bx",
				"wBSMl47uQib",
				"alvBxSxuorK",
				"SqN2upmg2Im",
				"SqJAH6Pu3Cs",
				"Ai48B5k6hwb",
				"M05iAqd6ZFU",
				"PWCUb3Se1Ie",
				"IXuqhJzEYP9",
				"I1Tn68LFq5Y"
			]
		},
		"HWF": {
			"dataSets": [
				"yMoTWVrYvwM",
				"Tu81BTLUuCT",
				"Humg4HbkmJg",
				"Wxl8hxuOsIQ",
				"m5MiTPdlK17",
				"pZ3XRBi9gYE",
				"HtZb6Cg7TXo",
				"cxfAcMbSZe1",
				"WDyQKfAvY3V",
				"ZRsZdd2AvAR",
				"p5z7F51v1ag",
				"LaD0zShIosR"
			]
		}
	},
	"preprocess": [{
			"selectDepartament": "NTD",
			"selectMetadataType": ["trackerPrograms"],
			"anonymizePhone": true,
			"anonymizeMail": true,
			"anonymizeOrgUnit": true,
			"anonymizeCoordinates": true,
			"selectTrackedEntityAttributes": ["oTvXfEywjT3", "n8E6WIyAwcC", "DwZNiXy5Daz", "FHw1NKy0PWY", "eQtZaLIO3XU",
				"na3ZJRtjpGH", "HkBG3DVELBM", "sKBh0kazOCk", "AAkZm4ZxFw7", "ENRjVGxVL6l", "aBaYLJryaMr", "iy884aJfYTc"
			],
			"action": "anonymizeData"
		},
		{
			"selectDepartament": "HWF",
			"selectMetadataType": ["datasets"],
			"action": "anonymizeData"
		}
	],
	"postprocess": [{
			"selectFromGroups": [
				"ADMIN reactivate after cloning"
			],
			"action": "activate"
		}
	]
}
```
   
### :art: Screenshots

### :bookmark_tabs: Others

Commit:
Refactor code

Commit:
fix typo

idelcano committed:
wip

idelcano committed:
work in progress

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
last changes in anonimizer

Commit:
revert commented code used by test in local

Commit:
remove wrong files added by error

Commit:
remove wrong files added by error

Commit:
remove unnused methods

Commit:
fix typo

Commit:
Merge branch 'development' into feature/anonimizer

Commit:
Merge branch 'feature/password_as_params' into feature/anonimizer

Commit:
remove all users and move admin roles to one existing user

Commit:
added example rules

Commit:
added new example params

Commit:
Added pre process to remove all confidential data and passwords

Commit:
Added post-process to fakerize user attributes

idelcano commented:
Added new config parameters:

Preprocess:

  {
    "selectDepartament": "ALL",
    "selectAdminUser": "eyeseetea",
    "excludeUsernames": [
      "example1", "example2", "example3"
    ],
    "selectOldAdminUser": "adminuserq",
    "action": "anonymizeUsers"
  },

Its executed using sql queries.

Post-process:


  {
    "selectUsernames": "*",
    "excludeUsernames": [
      "example1","example2"
    ],
    "action": "fakerizeAllUserFields"
  },
executed using d2apì

idelcano commented:
To remove all the tracker in 2.36 we mustat add:

delete from trackedentityprogramowner where programid in (select programid from program where uid in ('cn4tDPLFfOT'));
delete from trackedentityinstance where trackedentitytypeid = '135';

tokland commented on src/postprocess/apirequests.py:
```diff
@@ -40,6 +37,23 @@ def activate(api, users):
+    for user in users:
+        if user["userCredentials"]["username"] not in excludeUsers:
+            name = fake.name()
+            user["surname"] = name.split()[1]
```
[minor] [proposal] `name, surname = fake.name().split()`

tokland commented on src/postprocess/apirequests.py:
```diff
@@ -101,14 +115,24 @@ def get_users_by_usernames(api, usernames):
-        },
-    )
+    contains = lambda x: x in usernames
+    if contains("*"):
```
[very minor] Not sure this lambda contributes much.

tokland commented on src/preprocess/preprocess.py:
```diff
@@ -49,66 +72,117 @@ def remove_all(list_uid, f):
+                                                               anonimize_org_units, anonimize_phone, anonimize_mail,
+                                                               anonimize_coordinate,
+                                                               departament[key][program_type], f)
+                            if key_type == dataset_type:
```
[minor] When disjoint comparisons, `elif` is more declarative.

idelcano replied:
done

tokland commented on src/preprocess/preprocess.py:
```diff
@@ -49,66 +72,117 @@ def remove_all(list_uid, f):
+def get_rule_content(rule, rule_type):
+    if rule_type in rule.keys():
+        return rule[rule_type]
+    if rule_type in [anonymize_phone_rule, anonymize_coordinates_rule, anonymize_mail_rule]:
```
`elif`

idelcano replied:
done

tokland commented on src/preprocess/preprocess.py:
```diff
@@ -49,66 +72,117 @@ def remove_all(list_uid, f):
+    for rule_type in rule[metadata_type]:
+        if rule_type == check_type:
+            return True
+    return False
```
[minor] Check https://docs.python.org/3/library/functions.html#any

idelcano replied:
mm that code is not evaluating if the rule_type exist, is evaluating if the check_type exist in the metadata_type rule. 
But i think the actual code is wrong i have changed it. I need test it before ask for a new review

tokland commented on src/preprocess/query_generator.py:
```diff
@@ -151,81 +174,420 @@ def generate_delete_datasets_rules(datasets, data_elements, org_units, org_unit_
+    sql_org_units = convert_to_sql_format(org_units)
+    where_dataelements = ""
+
+    if sql_datasets_uids != "":
```
[minor] https://stackoverflow.com/a/2802748/188031

tokland commented on src/preprocess/query_generator.py:
```diff
@@ -151,81 +174,420 @@ def generate_delete_datasets_rules(datasets, data_elements, org_units, org_unit_
+                write(f, "UPDATE programstageinstance SET    eventdatavalues = jsonb_set(eventdatavalues, "
+                         "'{" + uid + ",value}', concat(\'\',concat(concat(\'" + "\"" + "\',(select concat(\'Redacted:\',"
+                         "round(random()*dataelementid+programstageinstanceid))"
+                         ' FROM   dataelement WHERE  uid = \'sPadHOO4SQY\'),\'' + '\"' + '\'),\'\'))::jsonb)'
```
This escaping can be hard to follow.`""" multi-line here, with free ' " quotes """`  should also work

tokland commented on src/preprocess/query_generator.py:
```diff
@@ -151,81 +174,420 @@ def generate_delete_datasets_rules(datasets, data_elements, org_units, org_unit_
+                         ' AND de.dataelementid IN (SELECT psde.dataelementid'
+                         ' FROM program p INNER JOIN programstage ps ON p.programid = ps.programid INNER JOIN'
+                         ' programstagedataelement psde ON psde.programstageid = ps.programstageid WHERE  p.uid IN (\'' + "', '".join(
+                    all_uid) + '\')) and de.uid = \'' + uid + '\'));')
```
[proposal] `"""SOME SQL HERE WITH VARS {x} {y}""".format(x=1, y=2)`

tokland reviewed (approved):
Code looks good, just some comments as notes (no need to change). I've focused on the Python code, I assume the complex anonymizing SQL code fulfills your needs. 

Let me know if I should take the time to test with a real database and check field by field.

idelcano commented:
@tokland many minor changes of line format, maybe could be merge first other pr and then make this changes to avoid have a lot of merge conflicts?

tokland commented:
Sure!

idelcano committed:
Merge branch 'development' into feature/anonimizer

idelcano committed:
requested changes

idelcano committed:
remove sapces

idelcano committed:
fix typo

idelcano committed:
fix typo

idelcano committed:
added control version

idelcano committed:
added control version and change format and fix some typo

idelcano commented:
@tokland Nacho asked me to continue with the changes, and I have updated the branch and added a new parameter to the configuration file to differentiate between 2.34 and 2.36 since some table has already changed.

ifoche reviewed (approved):
thanks @idelcano
